src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ActiveShardCountBasedShardSyncStrategyDecider.java

```diff
@@ -0,0 +1,226 @@
+    }
+
+    private PeriodicShardSyncScheduler getShardSyncScheduler() {
+        return new PeriodicShardSyncScheduler(kinesisProxy,
+            leaseManager,
+            config.getInitialPositionInStreamExtended(),
+            config,
+            new ScheduledThreadPoolExecutor(PERIODIC_SHARD_SYNC_THREAD_COUNT),
+            shardSyncer,
+            getLeaderPoller(),
+            workerId);
+    }
+
+    private PeriodicShardSyncStrategy createPeriodicShardSyncStrategy() {
+        return new PeriodicShardSyncStrategy(PeriodicShardSyncManager.getBuilder()
+                                                 .withWorkerId(workerId)
+                                                 .withLeaderPoller(getLeaderPoller())
+                                                 .withPeriodicShardSyncScheduler(getShardSyncScheduler())
+                                                 .build());
+    }
+
+
+    private ShardEndShardSyncStrategy createShardEndShardSyncStrategy() {
+        return new ShardEndShardSyncStrategy(new ShardSyncTaskManager(kinesisProxy,
+            leaseManager,
+            config.getInitialPositionInStreamExtended(),
+            config.shouldCleanupLeasesUponShardCompletion(),
+            config.shouldIgnoreUnexpectedChildShards(),
+            config.getShardSyncIntervalMillis(),
+            metricsFactory,
+            executorService,
+            shardSyncer));
+    }
+
+    /**
+     * Gets the leases and filters out the SHARD_END leases to get the count of
+     * active leases
+     *
+     * @return Active leases count
+     * @throws DependencyException
+     * @throws InvalidStateException
+     * @throws ProvisionedThroughputException
+     */
+    private int getActiveShardleasesCount()
+            throws DependencyException, InvalidStateException, ProvisionedThroughputException {
+        // double checked locking to ensure only a single thread populates the
+        // leases as all threads will get roughly the same view of the leases at
+        // startup. This keeps a check on IOPS utilisation of Leases table. Also
+        // once populated, the strategy will never move back from Periodic to
+        // Shard sync as the number of shards never reduce. For the other case
+        // of increasing shards, its a fair assumption that there would be at
+        // least one deployment and the workers would initialize again before
+        // that happens
+        if(leases == null) {
+            synchronized (LOCK) {
+                if (leases == null) {
+                    getLeases();
+                }
+            }
+        }
+        return leases == null ? 0
+                       : (int) leases.stream().filter(lease -> lease.getCheckpoint() != null && lease.getCheckpoint() != ExtendedSequenceNumber.SHARD_END)
+                                     .count();
+    }
+
+    private void getLeases() throws DependencyException, InvalidStateException, ProvisionedThroughputException {
```

Comment:
That would imply only one worker could be updating the leases at any given time. We just want that the leasesLastFetchTime should be accessed in a thread safe manner with multiple threads updating the leases concurrently

Reply:
You always want to update and read the two fields atomically. Otherwise, there is a risk of the data field corresponding to a different value of the timestamp field. If latency is a concern, perhaps the setting of the field (and not the scanning of leases table) could be made synchronized to minimize blocking.